### DISCUSSION TITLE
vcpkg + Github Actions + CMake presets

### DISCUSSION THREAD

**Discussion by alessandrostranieri** [microsoft/vcpkg#21730] [Q&A] [answered]
Hi! I am enjoying a lot using **vcpkg** but I am completely stuck creating a nice simple workflow for Github Actions. Mind you: I have just started to learn all 3 technologies in the subject (divide et impera, right?).

On my local machine I have installed vcpkg and exported `VCPKG_ROOT` to point to its directory. With this setup I am able to configure and build with CMake by using a preset that (among other things) contains this line:

```json
"CMAKE_TOOLCHAIN_FILE": {
  "value": "$env{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake",
  "type": "FILEPATH"
}
```

This works **beautifully** on both **Powershell** and **Ubuntu over WSL**. 

Then I tried to achieve the same with a Github workflow (currently only Ubuntu). The Github documentation reports that both Windows and Ubuntu have vcpkg available and its installation is available through the environment variable `VCPKG_INSTALLATION_ROOT` ([Windows](https://github.com/actions/virtual-environments/blob/main/images/win/Windows2022-Readme.md) and [Ubuntu](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md)). 

Since I didn't want to change my preset, I thought I could just redefine an environment variable in the workflow with something like this:

```yaml
env:
  VCPKG_ROOT: ${{ VCPKG_INSTALLATION_ROOT }}
```

I tried all possible syntax forms on the right-hand side of the assignment, but this didn't work: `VCPKG_ROOT` always ended up being empty.

Then I changed the toolchain line of the preset file to look like that:

```json
"CMAKE_TOOLCHAIN_FILE": {
  "value": "$env{VCPKG_INSTALLATION_ROOT}/scripts/buildsystems/vcpkg.cmake",
  "type": "FILEPATH"
}
```

This also didn't work.

Finally, the only way I have manage to have the workflow succeeding was to "hard-code" vcpkg's path on the runner with:

```yaml
env:
  VCPKG_ROOT: "/usr/local/share/vcpkg"
```

But this is not ideal, because I would like to set up a matrix to test on both Ubuntu and Windows.

My question would then be: 

What's the simplest way to use vcpkg with presets and Github Actions?

Thank you! 🖖

**Comment by m-kuhn** [comment, marked as answer]
You will probably have a vcpkg submodule in your repository, why not use that?

**Reply by alessandrostranieri** [reply]
That is my current approach, and it works quite fine. However, I was curious about how to leverage vcpkg that is already installed on github's runners.

**Comment by mathisloge** [comment]
You can always combine presets and "normal" program arguments. 
So `cmake -DCMAKE_TOOLCHAIN_FILE="${{ env.VCPKG_INSTALLATION_ROOT }}scripts/buildsystems/vcpkg.cmake" --preset ci`

**Reply by alessandrostranieri** [reply]
Yes, that could work, though embedding this information already in the preset seems so nice and convenient.

**Comment by dg0yt** [comment]
> The Github documentation reports that both Windows and Ubuntu have vcpkg available

> I was curious about how to leverage vcpkg that is already installed on github's runners.

AFAIK this is always some "historic" version. It will be updated when the agents are updated, i.e. your build may break when you don't expect it. Possibly mitigated by versioning in your project's manifest.

**Reply by alessandrostranieri** [reply]
I see, I didn't know that. Then probably adding vcpkg as a git submodule is currently the best solution.
